wallet/wallet.go

```diff
@@ -2373,6 +2373,174 @@ func (w *Wallet) GetTransactions(startBlock, endBlock *BlockIdentifier,
+	// the database.
+	_, bestHeight, err := chainClient.GetBestBlock()
+	if err != nil {
+		return nil, false, errors.New("cannot get best block")
+	}
+
+	// Get the transaction information from directly calling the backend
+	// endpoint.
+	var backendErr error
+	var txResult *btcjson.TxRawResult
+	switch client := chainClient.(type) {
+	case *chain.RPCClient:
+		txResult, err = client.GetRawTransactionVerbose(txHash)
+		if err != nil {
+			// If the backend returned an error we will still try to
+			// find the transaction in the database, but return the
+			// error if it was also not found there.
+			backendErr = err
+			bestHeight = 0
+		}
+	case *chain.BitcoindClient:
+		txResult, err = client.GetRawTransactionVerbose(txHash)
+		if err != nil {
+			// If the backend returned an error we will still try to
+			// find the transaction in the database, but return the
+			// error if it was also not found there.
+			backendErr = err
+			bestHeight = 0
+		}
+	case *chain.NeutrinoClient:
+		return nil, false, errors.New("not supported with neutrino client")
+	}
+
+	// Set the block hash from the provided string and get the block height
+	// from it.
+	//
+	// TODO: Fetching block heights by their hashes is inherently racy
+	// because not all block headers are saved but when they are for SPV the
+	// db can be queried directly without this.
+	var blockHash *chainhash.Hash
+	var endHeight int32 = -1
+	if txResult != nil && txResult.BlockHash != "" {
+		blockHash, err = chainhash.NewHashFromStr(txResult.BlockHash)
+		if err != nil {
+			return nil, false, err
+		}
+
+		// Obtain the block height from the backend.
+		switch client := chainClient.(type) {
+		case *chain.RPCClient:
+			header, err := client.GetBlockHeaderVerbose(blockHash)
+			if err != nil {
+				return nil, false, err
+			}
+			bestHeight = header.Height
+		case *chain.BitcoindClient:
+			bestHeight, err = client.GetBlockHeight(blockHash)
+			if err != nil {
+				return nil, false, err
+			}
+		}
+
+		// We know that the transaction definitively is in a specific
+		// block height and set both start and end height to it.
+		endHeight = bestHeight
+	}
+
+	// Populate with additional data if this transaction exists in the
+	// database.
+	var summary *TransactionSummary
+	var inDB bool
+	err = walletdb.View(w.db, func(dbtx walletdb.ReadTx) error {
+		txmgrNs := dbtx.ReadBucket(wtxmgrNamespaceKey)
+
+		rangeFn := func(details []wtxmgr.TxDetails) (bool, error) {
+			for i, txDetail := range details {
+				txDetail := txDetail
+				dbTx := makeTxSummary(dbtx, w, &txDetail)
+
+				// We are only interested in the specific
+				// transaction and can return early.
+				if txHash.IsEqual(dbTx.Hash) {
+					summary, inDB = &dbTx, true
+					if details[i].Block.Height != -1 {
+						bestHeight = details[i].Block.Height
+						blockHash = &details[i].Block.Hash
+					}
+					return true, nil
+				}
+			}
+			return false, nil
+		}
+
+		return w.TxStore.RangeTransactions(
```

Comment:
Why range over transactions vs looking up the specific-one we want? See the method I linked above

Reply:
I have changed the code to make use of `Store.TxDetails` by first checking if the transaction exists in the store, and only if not make use of the backend RPC.